### ISSUE TITLE
View milestones in /vote gui

### ISSUE THREAD

**Issue by TomLewis** [Ben12345rocks/VotingPlugin#1396]
Can we please get a way to check milestones in-game as a player, via maybe /vote gui, so that they can see how far they have come and to track how many more they can do.

Also when I add new vote milestones, are these atomatically given to players that supassed the milestones already? How/when are these milestones being triggerd when the vote?

You could add lore to the milestone yaml so thats shown as the lore:

```yaml
  '15000':
    Enabled: true
    DisplayName: '&a&lVote 21'
    DisplayLore:
    - '&7You have voted 21 times!'
    - '&7Click to receive your reward!'
    Rewards:
      Money: 15000
      Messages:
        Player: '&8[&3VOTE&8] &aYou got %milestone% milestone votes!'
      Commands:
        Console:
        - crate give to %player% vote 21
```

**Comment by BenCodez** [comment]
These placeholderapi placeholders are available

    votingplugin_milestonecount - User milestonecount
    votingplugin_milestone_numberofvotesuntil_20 - Get number of votes until milestone
    votingplugin_milestone_numberofvotesuntil_1 - Get number of votes until milestone
    votingplugin_nextmilestone_votes_required - Get number of votes required for next available milestone
    votingplugin_nextmilestone_votes_until - Get number of votes until next available milestone

**Comment by TomLewis** [comment]
![2024-11-23_22 42 41](https://github.com/user-attachments/assets/760f3139-f8d8-44ff-8227-1f88358e80e8)

I started making a custom GUI for this out of those placeholders, but theres an issue. The first time you load these placeholders, they do not work. It shows some dots.

This happens every single time you log out and in-again, which means if I have a dynamic Deluxe Menus that shows custom blocks based on if you have unlocked a milestone or not, it will show incorrectly with nothing, you would have to as a user open the GUI, close it and open it again for these to load...

If I close the panel and re-open it, its fine, until I relog.
![2024-11-23_22 45 18](https://github.com/user-attachments/assets/7c99e6dc-ec1f-4a49-bbc2-8cb00170c2f4)

**Comment by BenCodez** [comment]
https://github.com/BenCodez/VotingPlugin/wiki/PlaceHolderAPI-Expansion

Add _process to the end of the placeholder.

After a server restart it should stay cached though without that

**Comment by TomLewis** [comment]
Fantastic, that worked without having to restart, doing this all on my local dev server atm.

Are there any performance issues with this?

**Comment by BenCodez** [comment]
It just might trigger a mysql query, so try to avoid using that on the main thread, it can crash the server in rare cases. That's why it's disabled by default

**Comment by TomLewis** [comment]
How would I know if its on the main thread? Im using deluxe menus to make

**Comment by BenCodez** [comment]
From a quick look at the source code it might not be on the main thread (when pulling placeholders), but not certain though

**Comment by TomLewis** [comment]
Right, ive managed to make my GUI, I got it all working apart from one thing, I want to show progress bars using;
https://github.com/aBooDyy/Progress-Expansion

But with the placeholders we have, we can only make backwards placeholders with manually putting in "100" as thats the differnace between that and the next/previous milestones.
![2024-11-24_16 59 48](https://github.com/user-attachments/assets/46007a72-1adf-47b7-8ed3-bfab15c867f4)

Using
```yaml
    - '%progress_bar_{votingplugin_milestone_numberofvotesuntil_600_process}_c:&a■_p:&e■_r:&7■_l:10_m:100_fullbar:&aCompleted!%'
```

or I can use
```yaml
    - '%progress_bar_{votingplugin_milestonecount_process}_c:&a■_p:&e■_r:&7■_l:10_m:600_fullbar:&aCompleted!%'
 ```

Which isnt right either, as this displays from 0-600 progress, so if im doing a milestone thats 500-600 thats only 100 it needs to progress, at least its a progress bar in the right way.

we need placeholders that are not `numberofvotesuntil` but counts in the other way, so instead of the 90 until, it would be `numberofvotessofar` number of votes so far in the milestone, so it would be 10, this would allow us to use the progress bar!

For now, im going to have to use:
```yaml
    - '%progress_bar_{votingplugin_milestone_numberofvotesuntil_600_process}_c:&a■_p:&e■_r:&7■_l:10_m:100_fullbar:&aCompleted!%'
```

but, once ive finished my 500 milestone and move into the active 600 it already says in 3/4 done, because it counts from 0.

I tried using math placeholder to solve it, but I cant have a placeholder in a placeholder in a placeholder, deluemenu breaks it and sets it to 0.
```yaml
    - '%progress_bar_{math_0:0_100-{votingplugin_milestone_numberofvotesuntil_600_process}}_c:&a■_p:&e■_r:&7■_l:10_m:100_fullbar:&aCompleted!%'
``` 
This would be right, if it worked. where ive done `100-votingplugin_milestone_numberofvotesuntil_600_process}`  as 100 is the differanece between the 600-700 which is the next milestone.

**Comment by TomLewis** [comment]
Tldr of that, "current" active milestone amount of votes.

So if Ive got a 600-700 milestone, that's a total of 100 votes needed, 610 votes would be 10 current votes for that milestone.

**Comment by BenCodez** [comment]
https://bencodez.com/job/VotingPlugin/

%votingplugin_lastmilestone_votes_since% 

Does this work?

**Comment by TomLewis** [comment]
Thanks, but sadly this crashes my server with out of memory, I will try bumping the memory and re-testing.

**Comment by BenCodez** [comment]
Shouldn't be using that much. I did a quick test so I know it works somewhat

**Comment by TomLewis** [comment]
Very odd, I gave the server an extra Gig as its only a local test server so theres nothing running on this world at all and it now boots, but its now using a whole extra gig of memory for the first few mins, then it settles down again.

I can backdate and check again if you want

**Comment by BenCodez** [comment]
Do a memory dump or something along those lines and see what's using it if you can

**Comment by TomLewis** [comment]
yeah I backdated to version 6.17.3 and it does seem to load a lot of mem, but it dousnt crash the server on boot with the low memory, ill install spark and see what that says with the latest dev.

**Comment by BenCodez** [comment]
Could be the skull caching doing it then

**Comment by TomLewis** [comment]
Ive had half a million players join my server and I copied my data from my prod to my local test server, so if its loading all these players to do something... yeah that might do something to memory haha

**Comment by BenCodez** [comment]
Yeah that would do it. Previous versions of skull caching won't load if memory is low, may have to add something for that as well for the new version of it

**Comment by TomLewis** [comment]
ok second boot it booted fine no issues, so I presume its the cache first time it boots? I also noticed the database is MUCH larger. You may not want to do the cache for all players at once at boot, if theres a lot of players.

Well at least we found this now on a test server! Woohoo for accidental bug testing LOL

**Comment by BenCodez** [comment]
It loads the cache on boot, but very slowly, default one head every 3 seconds, only what's on /vote top

**Comment by BenCodez** [comment]
My guess the main ram usage might actually be the placeholders caching, so that happens all at once for the most part

**Comment by TomLewis** [comment]
![2024-11-24_22 58 32](https://github.com/user-attachments/assets/1e1c90e8-fb59-4a1e-a167-5fcb50af62be)
![2024-11-24_22 59 32](https://github.com/user-attachments/assets/55707229-79ff-456e-a6f1-ec097a3543c2)

As for the new placeholder, it seems to work exactly as intended! showing here 510/600 & 580/600.

Now im going to worry about this cache/ram.

im so tired I cant remember how to make a ram report with Spark, I might go to bed now bit late for me.
But I can carry on tomorrow if you want me to do a report.

But there's definitely a lot more ram being used in this newer version. its sitting on 2.6 GB as I bumped it to 3gb, its usually fine running on 2GB as its a local test server with nothing going on it.

Thanks so far!

**Comment by BenCodez** [comment]
Send a report whenever you can and I'll see what I can do about it. Does it happen with a fresh/smaller database by chance?

Could you also private message a copy of your entire database so I can test it?

**Comment by TomLewis** [comment]
Remind me how to make a report 🤣
It's been a while and I can't remember how.

Yeah I can send you a copy, it's in mysql format, hope that's fine. Have I got you on discord? My username is "GodsDead".

I'm sure I've sent data in the past.

It will have to be tomorrow mind you, I'm in bed now haha

**Comment by BenCodez** [comment]
Can also just create a ticket and send it there, just link back to here so I know what's it's for. Mysql dump or something along those lines should work fine

Don't exact remember how to do a report either, been awhile

**Comment by TomLewis** [comment]
I asked chatGPT and it said to use Spark and https://eclipse.dev/mat/

This is what it said:

Heap Dump Analysis

1.     Run /spark heap to generate a heap dump.
2.     Spark will save a .hprof file in your server's root directory.
3.     Analyze the .hprof file using a memory analyzer tool like:

This method shows detailed memory usage by objects, allowing you to see which plugins are using the most memory.


So ive done this and using eclipse MAT, its now parsing the file, but it seems to be taking ages... then Ill have to figure out how to read what it spits out

**Comment by BenCodez** [comment]
Are you able to send me it? Should also be able to use this: https://spark.lucko.me/

**Comment by TomLewis** [comment]
Yeah man, Not public mind you! Want to drop me a DM and il send you both things

**Comment by BenCodez** [comment]
You can message me on spigot or discord, or create a ticket on discord (tickets are not public)
